src/Database/Traits/PathEnumerable.php

```diff
@@ -0,0 +1,322 @@
+
+            $model->bindEvent('model.afterUpdate', function () use ($model) {
+                $model->moveToNewParent();
+            });
+
+            $model->bindEvent('model.beforeDelete', function () use ($model) {
+                $model->deleteDescendants();
+            });
+
+            if (static::hasGlobalScope(SoftDeletingScope::class)) {
+                $model->bindEvent('model.afterRestore', function () use ($model) {
+                    $model->restoreDescendants();
+                    $model->setEnumerablePath();
+                });
+            }
+        });
+    }
+
+    /**
+     * Gets the direct parent of the current record.
+     */
+    public function getParent(): Collection
+    {
+        return $this->parent()->get();
+    }
+
+    /**
+     * Gets all ancestral records of the current record.
+     */
+    public function getParents(): Collection
+    {
+        return $this->newQuery()->ancestors()->get();
+    }
+
+    /**
+     * Gets all direct children of the current record.
+     */
+    public function getChildren(): Collection
+    {
+        return $this->children()->get();
+    }
+
+    /**
+     * Gets all children (ancestors) of the current record.
+     *
+     * This will include children records of the child records, and so on.
+     */
+    public function getAllChildren(): Collection
+    {
+        return $this->newQuery()->descendants()->get();
+    }
+
+    /**
+     * Gets a nested collection of all records.
+     */
+    public function getNested(): Collection
+    {
+        return $this->newQuery()->get()->toNested();
+    }
+
+    /**
+     * Root nodes scope.
+     *
+     * Gets all record that form the root nodes of the hierarchy.
+     */
+    public function scopeRoot(Builder $query): void
+    {
+        $query->whereNull($this->getParentColumnName());
+    }
+
+    /**
+     * Descendants scope.
+     *
+     * Gets all children records, and all children of those records, and so on.
+     */
+    public function scopeDescendants(Builder $query): void
+    {
+        if (!$this->exists()) {
+            return;
+        }
+
+        $query->where($this->getPathColumnName(), 'LIKE', $this->getPath() . '/%');
+    }
+
+    /**
+     * Ancestors scope.
+     *
+     * Gets all records that are direct ancestors (parents) of the current record.
+     */
+    public function scopeAncestors(Builder $query): void
+    {
+        if (!$this->exists()) {
+            return;
+        }
+
+        $query->whereIn($this->getKeyName(), $this->getAncestorIds());
+    }
+
+    /**
+     * Gets the enumerable path on the current record.
+     *
+     * This will take into account any parent changes, allowing you to get the new path before the record is saved.
+     */
+    public function getEnumerablePath(): string
+    {
+        if ($this->parent()->exists()) {
+            return $this->parent->{$this->getPathColumnName()} . '/' . $this->id;
+        }
+
+        return '/' . $this->id;
```

Comment:
@bennothommo Does this have to be hardcoded to reference the `id` column? What if I wanted to use a custom column name, say `folder` *cough cough*

Reply:
We can make a constant for the heathens.

Reply:
What are the practical reasons for not wanting to use the actual folder names for the Media Library use case?

Reply:
@LukeTowers probably nothing in terms of queries and functionality, but it could potentially result in larger path strings which would mean less levels of directories before you hit MySQL's string limits. At least with IDs you'll have a lot more levels before hitting those limits.

Reply:
Fair enough, do you think it would be worth adding in a constant for the max length of the path to surface that before it causes issues?

Reply:
@LukeTowers I would say if the intention is to try and catch it before it becomes a problem, it's probably worth querying the schema directly and working out the maximum length, as opposed to an arbitrary max length constant. But either way, it will mean that the "segment" column (id/folder/etc.) would need to be configurable in some fashion, so yeah, it's either a constant or a method to get that column.

Reply:
Further to this, I opted not to try and find the maximum length of the path column in the schema as the only way to do so is to run an SQL query, which means it would call this every page load that a enumerable path model is used (unless it were cached).

I think this is something we can opt to document as opposed to trying to technically handle - most people, I imagine, would opt to use the ID paths, and for people who know how to change the segment column, they (should?) be knowledgeable enough to ensure their DB schema accounts for the potentially larger path values.